### PR TITLE
Effective Programmers Read Code

### PR DESCRIPTION
**Pull request by bbengfort**
A quick essay about reading code to become a more effective programmer. 

I'd appreciate any help to make this as solid an essay as possible. In particular I'm not necessarily a fan of the title or the featured image. Any and all revisions are welcome to make this sharper and more effective of a post!

### CODE REVIEW

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+slug: "effective-programmers-read-code"
+date: "2022-08-05T08:10:14-04:00"
+draft: true
+image_webp: images/blog/reading.webp
+image: images/blog/reading.jpg
+author: Benjamin Bengfort
+description: "While we often think ourselves as authors of code we probably spend more time reading code than writing it. Changing our mindset to think of ourselves as avid code readers will make us more effective programmers."
+---
+
+Programmers think of themselves as authors of code but our daily engineering practice probably requires us to spend more time reading code than writing it. Slightly changing our mindset to think of ourselves as avid code readers will make us more effective programmers and ensure that we can quickly create mental maps of our code, a technique that I refer to as the "glass castle".
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:12]: Add comma after code?

Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:12]: Suggestion: As programmers, we tend to think of ourselves as authors of code, but our daily engineering practice...

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+<!--more-->
+
+When we write source code, we write for two audiences: the execution runtime that the code operates in and other programmers that need to use or maintain the code including future you. Most discussions about how to write good code focus on one or the other of these audiences, however there is much more middle ground between these audiences than those discussions suggest. Both the runtime and the reader have to step through code line by line, interpreting it and possibly jumping to other locations in code to understand the next step.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:16]: Suggestions: "...other programmers that need to use or maintain the code - including future you."
"Most discussions about how to write good code focus on just one of these audiences, however there is..."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+The operations and jumps described by source code create complex interactions that define the dynamics of our software systems. The only difference between the runtime and the reader is that the runtime handles a long sequence of instructions and the reader builds a mental abstraction of the interactions. As we program, we build this abstraction in our heads and we must mentally hold it in its entirety while we program. Because of the fragility of the model, (even a minor distraction can bring it crashing down) I refer to it as the "glass castle".[^1]
+
+For me, the glass castle is both a structure and a map.[^2] I hold individual named components in mental space such that related components are nearer to each other. The components are linked with mental threads based on their interactions: e.g. calling a method or a function of another component or instantiating and referencing an object. When I mentally zoom into a component I see that it is itself made up of components with interactions to each other. As I code, I mentally zoom in and out of the model getting more detail when I need it and encapsulating and isolating detail when I don't.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:20]: Suggestions: "...in mental space in a way that places related components near one another."
"As I code, I mentally zoom in and out of the model, getting more detail when..."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+The primary purpose of the glass castle is to ensure that we can quickly locate related code in the codebase and understand the consequences of making changes to the code. It does serve another purpose however: like [code "smells"](https://martinfowler.com/bliki/CodeSmell.html), I use the glass castle to understand when there is a structural or architectural problem with the system. For example, when I zoom into a component and I see that there are too many external interactions with subcomponents, it's an indication that the parent component is not correctly abstracted. This is very similar to the idea of ["wide vs deep" modules](https://nakabonne.dev/posts/depth-of-module/) described in Ousterhout's [_A Philosophy of Software Design_](https://web.stanford.edu/~ouster/cgi-bin/book.php). The best use of the glass castle is to determine when the system is "tipping" over as we add features or make changes because it may lead us to refactoring or addressing the initial design that wasn't quite right for the current iteration of software.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:22]: Suggestions: "It does serve another purpose, however; like [code "smells"]..."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+The glass castle is essential for how I write good code and recently, I've started to pay more attention to how I construct the glass castle. When I sit down to implement a feature or make a change, the very first thing I do is identify candidate locations in the code where the change needs to be made. These locations are identified in a top down fashion looking at the organization of the code and quickly scanning through package and module names and package documentation.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:24]: Suggestions: "The glass castle is essential to my process of writing good code. Recently, I've started to pay more attention to..."
"These locations are identified in a top-down fashion: I look at the organization of the code and quickly scan through package and module names as well as package documentation.

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+Once I find the candidate locations, I continue in a bottom up fashion identifying the functional dependencies surrounding the location and putting together components and interactions tagged with the input and output data contracts that they make. Often, I have to read the code of the external dependencies of the project to figure out the best possible solutions. I continue grouping, linking, and encapsulating code in my head until I feel like I have enough of a picture that I'm confident in the change I'm about to make.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:26]: Suggestions: "...I continue in a bottom-up fashion by identifying the functional dependencies..."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+Before I've written even a single line of code, I've read hundreds of lines of code. The more complex the change or the system (or the longer I was away from the system before making a change), the more time I have to spend reading code before writing it. That's when I realized that as a programmer I spend far more time reading code than writing it.[^3]
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:28]: Suggestion: "Before I've written even a single line of code, I've already read hundreds of lines of code. The more complex the change or system (or, the longer I've been away from the system before making a change),..." 

"It's through situations like these that I've come to realize, as a programmer, I spend far more time reading code than writing it."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+My students often ask me how to become a better programmer and my answer was always "read code, particularly good code, and you'll become a better programmer". The theory was similar to writing blogs or books -- the more you read, the better you write and there is tons of excellent open source code. More importantly, I noticed that my students would ignore blocks of code to quickly get to writing code and more often than not, the ignored blocks were where the problems came from. What I understand now is that it's not just practice or a way to become better or avoid bugs. Instead, reading code is the most essential part of our job as programmers.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:30]: Suggestions: "...and my answer has always been "read code..."

"....the more you read, the better you'll write - and there are tons of resources for excellent open source code."

"...get to writing code and, more often than not, those ignored blocks were where the problems came from."

"it's not just practice, or a way to become better or avoid bugs."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+The more code we read the quicker and more effective we are at building the glass castle, the better code we will write. To become effective authors of good code, we should be first avid readers of good code.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:32]: Suggestion: "The more code we read, the quicker and more effective we are at building the glass castle, the better the code is that we will write."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+#### P.S. Tips for code readers
+
+Reading code must also be a practice that we focus on and we should prepare ourselves for success.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:36]: Suggestion: "Reading code must be a practice that we focus on in order to prepare ourselves for success."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+1. _Ensure you have long periods of uninterrupted development time._ Give yourself as much time with the glass castle as possible. At Rotational we try to schedule all of our meetings for Tuesdays and Thursdays to leave Monday and Wednesday for deep work.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:38]: Suggestion: "At Rotational, we try to schedule..."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+2. _Create shortcuts to build the glass castle faster._ If you can create mental checkpoints, notes to yourself, or other mnemonics you will avoid treading the same path over and over again.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:39]: Suggestion: insert comma after mnemonics

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+3. _Diversify the code you are reading._ Don't just read the code that you or your colleagues wrote, or even stick to code in the language you're working in. There are a huge number of excellent open source projects that you can read; make this practice part of your continuous learning strategy.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:40]: Suggestion: "There is a huge number of excellent..."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+4. _Glass castle before code review._ The glass castle strategy isn't just about implementing code, it's also about reviewing code.
+5. _Read code to write documentation._ It is difficult to make documentation part of the definition of done of a task, instead create tasks for documentation that allow you to read code from a user or test developers perspective.
+
+#### P.S. Writing for those who read code
+
+We started this essay by discussing the two audiences for code: the runtime and the reader. The runtime is the execution environment of the code including the compiler or the interpreter, other code that imports or loads your code, and the operating system that runs your code in a process. The reader is other programmers: users of your code as a module, developers who are writing tests for your code, maintainers who deploy code, fix bugs and improve performance, and future you who has the expertise to understand what, how, and why of the source code.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:46]: Suggestion: "...to understand the 'what', 'how', and 'why' of the source code.

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+
+The first and most important notion is that these audiences are themselves dynamic. The more time that passes from the original implementation of the source code, the more different both the runtime and the reader are. We should write code to ensure longevity of executability and understandability.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:48]: "...these audiences are, themselves, dynamic."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+When it comes to writing code, we certainly should not only focus on or prioritize execution performance if it sacrifices readability. Self documenting code is absolutely something to strive for. On the other hand, clean coders eschew comments and have probably have taken this to the extreme. Obtuse code serves no one, not future readers and not compilers. Verbose comments that repeat the code make it harder to build the glass castle. Instead, use comments to create markers or way points that help guide the reader to other pieces of important or related code, or that serve to build the glass castle more efficiently.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:50]: Suggestion: "On the other hand, some clean coders who comment in excess have probably have taken this to an extreme."

"code, or places that serve to build the glass castle more efficiently."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+
+Documentation is essential and should give us a high level overview of what the code should do. The best documentation gives us entry points into the code so that we can understand it further, but doesn't go so far that it prevents you from engaging with the code itself. Ensure that all packages, structs, methods, and functions are correctly documented with doc strings and that all are reachable from the map that the documentation describes.
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:52]: "engaging with the code, itself."

File: content/en/blog/2022-08-05-effective-programmers-read-code.md [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+[^2]: My experience of the glass castle is as a graph or molecular structure which I can "see" in my mind. In fact, sight is the best way to describe my interaction with the castle sort of like how reading instead of hearing is the best way to describe how I consume an audiobook. It's a bit hard to describe what I see, just as it would be hard to describe a dream, but I think the important note here is that as you program, you should be mentally interacting with this model in a way that makes sense to you, e.g. is it "touch" or "movement" for you?
+
+[^3]: And that is before code reviews!
```
Comment (nwhite222) [inline content/en/blog/2022-08-05-effective-programmers-read-code.md:59]: This was a really cool perspective and read! The suggestions I had were all grammar/clarity-related...the content looks great to me:)

Comment (nwhite222) [reply content/en/blog/2022-08-05-effective-programmers-read-code.md:59]: Per the image, are we still keeping with the 'nature' theme of previous blog images? If so, the one I'm attaching below might be an option? (an example of 'reading' nature
BN-WO094_bkrvna_M_20171212115450
)

Comment (rebeccabilbro) [reply content/en/blog/2022-08-05-effective-programmers-read-code.md]: We loved this image but realized we couldn't use it with the copyright permissions 😞, so I've updated the image to use another one with a similar (I hope) vibe 🌳

**Review by rebeccabilbro — APPROVED** [review]
Excellent post @bbengfort, it was a very enjoyable read. Thank you @nwhite222 for your thoughtful edits; I've incorporated them and done some additional light editing. I'll go ahead and get this published right away!

**Pull request merged**